lib/internal/modules/esm/assets/load.js

```diff
@@ -0,0 +1,300 @@
+              bodyStream.close();
+              dispose();
+              reject(e);
+              r(e);
+            };
+            res.on('error', onError);
+            res.pipe(bodyStream);
+          } else if (res.headers['content-encoding'] === 'gzip' ||
+            res.headers['content-encoding'] === 'deflate') {
+            bodyStream = createUnzip();
+            onError = function onError(e) {
+              bodyStream.close();
+              dispose();
+              reject(e);
+              r(e);
+            };
+            res.on('error', onError);
+            res.pipe(bodyStream);
+          } else {
+            onError = function onError(e) {
+              dispose();
+              reject(e);
+              r(e);
+            };
+          }
+          bodyStream.on('error', onError);
+          bodyStream.on('data', (d) => {
+            ArrayPrototypePush(buffers, d);
+            size += d.length;
+          });
+          bodyStream.on('end', () => {
+            const body = entry.body = /** @type {Buffer} */(
+              BufferConcat(buffers, size)
+            );
+            f(body);
+          });
+        }),
+      };
+      getCache.set(parsed.href, entry);
+      fulfill(entry);
+    });
+  });
+  getCache.set(parsed.href, result);
+  return result;
+}
+
+/**
+ * Returns if an address has local status by if it is going to a local
+ * interface or is an address resolved by DNS to be a local interface
+ * @param {string} hostname url.hostname to test
+ * @returns {Promise<boolean>}
+ */
+async function isLocalAddress(hostname) {
+  try {
+    if (StringPrototypeStartsWith(hostname, '[') &&
+        StringPrototypeEndsWith(hostname, ']')) {
+      hostname = StringPrototypeSlice(hostname, 1, -1);
+    }
+    const addr = await dnsLookup(hostname, { verbatim: true });
+    if (addr.family === 4) {
+      return addr.address === '127.0.0.1';
+    } else if (addr.family === 6) {
+      return normalizeIPv6(addr.address) ===
+        '0000:0000:0000:0000:0000:0000:0000:0001';
+    }
+  } catch {}
+  return false;
+}
+
+/**
+ * Used to expand interface addresses to a canonical form for comparisons
+ * E.G. comparing ::1, 0::1, etc. to 0000:0000:0000:0000:0000:0000:0000:0001
+ * Due to IPv6 having multiple ways to represent a single address we need to
+ * normalize
+ * @param {string} str
+ * @returns {string}
+ */
+function normalizeIPv6(str) {
+  const parts = StringPrototypeSplit(str, ':');
+  let expansionIndex = -1;
+  for (let i = 0; i < parts.length; i++) {
+    const part = parts[i];
+    if (part === '' && expansionIndex === -1 && parts.length < 8) {
+      expansionIndex = i;
+      continue;
+    }
+    parts[i] = StringPrototypePadStart(part, 4, '0');
+  }
+  if (parts.length < 8) {
+    ArrayPrototypeSplice(
+      parts,
+      expansionIndex,
+      1,
+      ...ArrayFrom({ length: 8 - parts.length + 1 }, () => '0000')
+    );
+  }
+  return ArrayPrototypeJoin(parts, ':');
+}
```

Comment:
Is it worth moving this to a separate utility somewhere? Seems like something that could work towards being a public function in due course?

Reply:
could be done but unclear where it should go, probably on `net` but could be done as a separate PR